src/UglyToad.PdfPig/Writer/PdfMerger.cs

```diff
@@ -286,79 +266,111 @@ private ObjectToken CopyPageNode(PageTreeNode pageNode, IndirectReferenceToken p
+                    {
+                        // This is because, since we don't write token directly to the stream. So can't know the offset.
+                        // The token would be invalid. Although I don't think the copy of an object token would ever happen
+                        throw new NotSupportedException("Copying a Object token is not supported");
+                    }
                 }
-                else if (tokenToCopy is IndirectReferenceToken referenceToken)
-                {
-                    var tokenObject = DirectObjectFinder.Get<IToken>(referenceToken.Data, tokenScanner);
-
-                    Debug.Assert(!(tokenObject is IndirectReferenceToken));
 
-                    var newToken = CopyToken(tokenObject, tokenScanner);
-                    var objToken = context.WriteObject(memory, newToken);
-                    return new IndirectReferenceToken(objToken.Number);
-                }
-                else if (tokenToCopy is StreamToken streamToken)
-                {
-                    var properties = CopyToken(streamToken.StreamDictionary, tokenScanner) as DictionaryToken;
-                    Debug.Assert(properties != null);
-                    return new StreamToken(properties, new List<byte>(streamToken.Data));
-                }
-                else // Non Complex Token - BooleanToken, NumericToken, NameToken, Etc...
-                {
+                if (!deepCopy)
                     return tokenToCopy;
-                }
-            }
 
-            private void WriteHeaderToStream()
-            {
-                WriteString($"%PDF-{currentVersion:0.0}", memory);
-
-                memory.WriteText("%");
-                memory.WriteByte(169);
-                memory.WriteByte(205);
-                memory.WriteByte(196);
-                memory.WriteByte(210);
-                memory.WriteNewLine();
-            }
+                // This tokens can't be deep copied
+                if (tokenToCopy is EndOfLineToken || tokenToCopy is NullToken || tokenToCopy is BooleanToken)
+                    return tokenToCopy;
+                
+                switch (tokenToCopy)
+                {
+                    case CommentToken commentToken:
+                        return new CommentToken(commentToken.Data);
 
-            private static void WriteString(string text, Stream stream)
-            {
-                var bytes = OtherEncodings.StringAsLatin1Bytes(text);
-                stream.Write(bytes, 0, bytes.Length);
-                stream.WriteNewLine();
+                    case HexToken hexToken:
+                        return new HexToken(hexToken.Data.ToCharArray());
+                    
+                    case InlineImageDataToken imageDataToken:
+                        return new InlineImageDataToken(imageDataToken.Data);
+                    
+                    case NameToken nameToken:
+                        return NameToken.Create(nameToken.Data);
```

Comment:
You don't actually need to deep copy the name token afaik, there's an inner cache which ensures only a single name token value for a given string can exist in the AppDomain.

Reply:
Thinking more about this I don't think we need this deep copy step. Things like NameToken, NumericToken, OperatorToken, etc are cached as far as possible so that we don't allocate too many instances. I think I'm misunderstanding the benefit/fix the cloning brings in this case.

Reply:
To be honest, I can't remember my reasoning behind implementing the deep copy. I guess it was something related to copy specific things from documents. But I guess I didn't take in consideration, that as of right know. Most of DataToken properties are `readonly` so there is no point.